src/smart_controller_ui.cpp

```diff
@@ -0,0 +1,179 @@
+    // print bars on the slider
+    mvwprintz( win, point( LEFT_MARGIN + lo_slider_x, y + 1 ), selection == 1 &&
+               slider == 0 ? hilite( white ) : c_red, "|" );
+    mvwprintz( win, point( LEFT_MARGIN + hi_slider_x, y + 1 ), selection == 1 &&
+               slider == 1 ? hilite( white ) : lgreen, "|" );
+
+    // user manual
+    y += MENU_ITEM_HEIGHT;
+    mvwprintz( win, point( LEFT_MARGIN, y ), selection == 2 ? hilite( white ) : gray,
+               _( "User manual" ) );
+
+    // key descriptions
+    std::string keys_text = string_format(
+                                _( "Use [<color_yellow>%s</color> and <color_yellow>%s</color>] to select option.\n"
+                                   "Use [<color_yellow>%s</color>] to change value.\n"
+                                   "Use [<color_yellow>%s</color> or <color_yellow>%s</color>] to switch between sliders.\n"
+                                   "Use [<color_yellow>%s</color> and <color_yellow>%s</color>] to move sliders."
+                                   "Use [<color_yellow>%s</color>] to apply changes and quit." ),
+                                input_ctx.get_desc( "UP" ),
+                                input_ctx.get_desc( "DOWN" ),
+                                input_ctx.get_desc( "CONFIRM" ),
+                                input_ctx.get_desc( "NEXT_TAB" ),
+                                input_ctx.get_desc( "CONFIRM" ),
+                                input_ctx.get_desc( "LEFT" ),
+                                input_ctx.get_desc( "RIGHT" ),
+                                input_ctx.get_desc( "QUIT" ) );
+
+    int keys_text_w =  WIDTH - 2;
+    int keys_text_lines_n = foldstring( keys_text, keys_text_w ).size();
+    fold_and_print( win, point( 1, HEIGHT - 1 - keys_text_lines_n ), keys_text_w, gray, keys_text );
+
+    wnoutrefresh( win );
+}
+
+void smart_controller_ui::control()
+{
+
+    ui_adaptor ui;
+    ui.on_screen_resize( [this]( ui_adaptor & ui ) {
+        win = init_window();
+        ui.position_from_window( win );
+    } );
+    ui.mark_resize();
+    ui.on_redraw( [this]( const ui_adaptor & ) {
+        refresh();
+    } );
+
+    std::string action;
+
+    do {
+        ui_manager::redraw();
+        action = input_ctx.handle_input();
+
+        if( action == "CONFIRM" || ( action == "NEXT_TAB" && selection == 1 ) ) {
+            switch( selection ) {
+                case 0:
+                    settings.enabled = !settings.enabled;
+                    break;
+                case 1:
+                    slider = 1 - slider;
+                    break;
+                case 2:
+                    const translation manual =
+                        SNIPPET.random_from_category( "smart_engine_controller_manual" ).value_or( translation() );
+
+                    scrollable_text( [&] {
+                        win = init_window();
+                        ui.position_from_window( win );
+                        return win;
```

Comment:
```suggestion
                        return init_window();
```
The resize callback of `ui` will be called when resizing, this would only unnecessarily repeat it.

Reply:
I've rechecked with the debugger, and you're right, `ui::resize` callback is called right before this. So it simplified it even further.